// react-dragable/lib/Dragable/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

require("./style.css");

var _constants = require("../utils/constants");

var _rotateResize = _interopRequireWildcard(require("../utils/rotateResize"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var DRAG_MOVING = _constants.DRAG_EVENT_MAP.DRAG_MOVING,
    DRAG_STOP = _constants.DRAG_EVENT_MAP.DRAG_STOP;

var RotateArrow = function RotateArrow(props) {
  var onMouseDown = props.onMouseDown;
  return _react.default.createElement("svg", {
    className: "archor-rotate",
    onMouseDown: onMouseDown,
    viewBox: "0 0 1024 1024",
    width: "20",
    height: "20"
  }, _react.default.createElement("path", {
    d: "M935.497143 659.017143a36.937143 36.937143 0 0 1-1.755429 18.505143 434.322286 434.322286 0 0 1-370.980571 337.773714C325.632 1045.650286 107.373714 876.251429 77.531429 637.293714a435.712 435.712 0 0 1 378.587428-487.131428c5.778286-1.243429 13.385143-1.901714 19.748572-1.974857-2.633143-24.137143-5.851429-46.518857-7.899429-66.267429-3.291429-25.526857-6.582857-44.617143-6.582857-50.980571-1.974857-7.021714-1.389714-15.36 3.657143-21.065143a21.796571 21.796571 0 0 1 31.305143-4.608l0.585142 1.243428 17.993143 16.530286 126.683429 98.962286 17.92 13.312 15.36 11.410285c5.12 3.803429 7.68 8.850286 8.996571 14.628572a20.260571 20.260571 0 0 1-4.388571 16.603428l-11.410286 15.36-13.312 17.92L555.885714 337.92l-14.555428 18.578286-1.243429 0.658285a23.625143 23.625143 0 0 1-31.890286 3.291429c-5.778286-5.12-9.654857-12.726857-8.411428-19.748571-1.316571-5.705143-3.291429-25.453714-6.582857-50.980572-2.633143-17.846857-5.266286-38.838857-7.899429-59.904-6.436571 0-13.385143 1.974857-19.748571 1.974857-193.828571 24.502857-332.726857 201.801143-308.297143 395.556572 24.502857 193.828571 201.801143 332.726857 395.556571 308.224a353.426286 353.426286 0 0 0 303.616-281.307429 40.082286 40.082286 0 0 1 34.377143-30.134857c22.235429-3.291429 43.373714 13.165714 44.763429 34.889143z"
  }));
};

var defaultProps = {};

var Dragable =
/*#__PURE__*/
function (_Component) {
  _inherits(Dragable, _Component);

  function Dragable() {
    var _this;

    _classCallCheck(this, Dragable);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Dragable).apply(this, arguments));
    _this.thisNode = null; // 便于计算的缓存变量

    _this.action = _rotateResize.Actions.NONE;
    _this.mouseStart = _rotateResize.EMPTY_POINT; // 变换开始前的鼠标位置

    _this.rectStart = _rotateResize.EMPTY_RECT; // 变换开始前的rect

    _this.fixedRatio = false;
    _this.archor = "s";
    /**
     * 组件拖拽（进行中处理）
     *
     * @type {EventListener}
     * @memberof Dragable
     */

    _this.handleDragMove = function (e) {
      var onChange = _this.props.onChange;

      var _assertThisInitialize = _assertThisInitialized(_this),
          mouseStart = _assertThisInitialize.mouseStart,
          rectStart = _assertThisInitialize.rectStart,
          action = _assertThisInitialize.action,
          archor = _assertThisInitialize.archor,
          fixedRatio = _assertThisInitialize.fixedRatio;

      var mouseEnd = (0, _rotateResize.getMousePoint)(e);
      onChange((0, _rotateResize.default)(action === _rotateResize.Actions.RESIZE ? archor : action, mouseStart, mouseEnd, rectStart, fixedRatio));
    };
    /**
     * 组件拖拽 (开始)
     *
     * @memberof Dragable
     */


    _this.handleDragStart = function (e, action) {
      e.stopPropagation();
      var event = e.nativeEvent;
      var rect = _this.props.rect;
      _this.mouseStart = (0, _rotateResize.getMousePoint)(event);
      _this.rectStart = rect;
      _this.action = action;
      _this.fixedRatio = e.shiftKey; // 绑定 鼠标滑动与释放事件

      (0, _utils.bindEvent)(document, DRAG_MOVING, _this.handleDragMove);
      (0, _utils.bindEvent)(document, DRAG_STOP, _this.handleDragStop); // 设置 背景鼠标友好提示

      switch (action) {
        case _rotateResize.Actions.MOVE:
          (0, _utils.setBodyCursor)("move");
          break;

        case _rotateResize.Actions.ROTATE:
          (0, _utils.setBodyCursor)("pointer");
          break;

        case _rotateResize.Actions.RESIZE:
          var cursor = event.target.style.cursor;
          (0, _utils.setBodyCursor)(cursor);
          _this.archor = cursor.replace(/-resize/, "");
          break;

        default:
          break;
      }
    };
    /**
     * 组件拖拽 (开始结束)
     *
     * @memberof Dragable
     */


    _this.handleDragStop = function () {
      //解绑拖拽事件
      (0, _utils.unBindEvent)(document, DRAG_MOVING, _this.handleDragMove);
      (0, _utils.unBindEvent)(document, DRAG_STOP, _this.handleDragStop); //重设

      (0, _utils.setBodyCursor)("auto");
    };

    return _this;
  }

  _createClass(Dragable, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      (0, _utils.unBindEvent)(document, DRAG_MOVING, this.handleDragMove);
      (0, _utils.unBindEvent)(document, DRAG_STOP, this.handleDragStop);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          rect = _this$props.rect,
          children = _this$props.children;
      return _react.default.createElement("div", {
        onMouseDown: function onMouseDown(e) {
          return _this2.handleDragStart(e, _rotateResize.Actions.MOVE);
        },
        className: (0, _classnames.default)("dragable", {
          "dragable-active": true
        }),
        style: {
          left: rect.x,
          top: rect.y,
          width: rect.width,
          height: rect.height,
          transform: "rotate(".concat(rect.rotate, "deg)")
        }
      }, _react.default.isValidElement(children) ? children : children(rect), _rotateResize.ARCHOR.map(function (direct) {
        return _react.default.createElement("i", {
          key: direct,
          className: (0, _classnames.default)("archor", "archor-".concat(direct)),
          style: {
            cursor: "".concat(direct, "-resize")
          },
          onMouseDown: function onMouseDown(e) {
            return _this2.handleDragStart(e, _rotateResize.Actions.RESIZE);
          }
        });
      }), _react.default.createElement(RotateArrow, {
        onMouseDown: function onMouseDown(e) {
          return _this2.handleDragStart(e, _rotateResize.Actions.ROTATE);
        }
      }));
    }
  }]);

  return Dragable;
}(_react.Component);

Dragable.defaultProps = defaultProps;
var _default = Dragable;
exports.default = _default;
